File: leetcode/42.接雨水.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l = [0] * n
        mx = 0
        for i in range(n):
            mx = max(mx, height[i])
            l[i] = mx
        mx = 0
        r = [0] * n
        for i in range(n-1, -1, -1):
            mx = max(mx, height[i])
            r[i] = mx
        ans = 0
        for i in range(n):
            ans += (max(0, min(l[i], r[i]) - height[i]))
        return ans
        

File: leetcode/42.接雨水.cpp
const int N = 20010;
int r[N];//维护右边最大值
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int mx = INT_MIN;
        for(int i = n-1; i >= 0; i--){
            mx = max(mx, height[i]);
            r[i] = mx;
        }
        mx = INT_MIN;
        int ans = 0;
        for(int i = 0; i < n; i++){
            mx = max(mx, height[i]);
            ans += (max(0, min(mx, r[i])-height[i]));
        }
        return ans;
    }
};